### PR TITLE
Provide a user-defined pruner example

### PR DESCRIPTION
**Pull request by tktran**
## Motivation
This PR addresses issue #1619. 

## Description of the changes
This PR consists of a Python file that defines a simple pruner that prunes a trial if it is in last place compared to all other trials, after a warmup period. It uses the pruner in a simple SGD hyperoptimization problem on the sklearn iris dataset. The top of the file contains a tutorial written in Sphinx markup. 

## TODO
I notice that, under docs/source/tutorial/, each tutorial is represented by 5 files: a .pickle, .ipynb, .py, .py.md5, and a .rst file - it looks like these are automatically generated somehow, but I couldn't figure out how, so I just wrote a .py in the meantime. I can make them (run some Sphinx command?) if shown.

## NOTES

Looking forward to other suggestions as well. Thank you -

### CODE REVIEW

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+An Example: Implementing LastPlacePruner
+--------------------------------------------------
+
+We aim to optimize the `loss` and `alpha` hyperparameters for a stochastic gradient descent classifier (SGDClassifier) run on the sklearn iris dataset. We implement a pruner which terminates a trial at a certain step if it is in last place compared to completed trials at the same step. We begin considering pruning after a "warmup" of 1 training step and 5 completed trials. For demonstration purposes, we print() a diagnostic message from prune() when it is about to return True (indicating pruning). 
+
+It may be important to note that the SGDClassifier score, as it is evaluated on a holdout set, decreases with enough training steps due to overfitting. This means that a trial could be pruned even if it had a favorable (high) value on a previous training set. After pruning, Optuna will take the intermediate value last reported as the value of the trial.
```
Comment (tktran) [inline tutorial/20_recipes/006_user_defined_pruner.py:24]: This line references a programming decision made at:
https://github.com/optuna/optuna/blob/35cc4c14471b501f1c76975bd587448fbe70c8ce/optuna/_optimize.py#L202

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 2]
```diff
@@ -4,24 +4,53 @@
...
+place compared to completed trials at the same step. We begin considering 
+pruning after a "warmup" of 1 training step and 5 completed trials. For 
+demonstration purposes, we print() a diagnostic message from prune() when it is 
+about to return True (indicating pruning). 
+
+It may be important to note that the SGDClassifier score, as it is evaluated on 
+a holdout set, decreases with enough training steps due to overfitting. This 
+means that a trial could be pruned even if it had a favorable (high) value on a 
+previous training set. After pruning, Optuna will take the intermediate value 
+last reported as the value of the trial.
```
Comment (tktran) [inline tutorial/20_recipes/006_user_defined_pruner.py:53]: This line references a programming decision made at:
https://github.com/optuna/optuna/blob/35cc4c14471b501f1c76975bd587448fbe70c8ce/optuna/_optimize.py#L202

**Comment by crcrpar** [comment]
Hi @tktran I'm super excited to see this pull request of a new tutorial

> I notice that, under docs/source/tutorial/, each tutorial is represented by 5 files: a .pickle, .ipynb, .py, .py.md5, and a .rst file - it looks like these are automatically generated somehow

Currently, our documentation incorporates Sphinx Gallery which runs python files put under the specified directory (in Optuna, `turoail` directory) and generates the corresponding rst and ipynb files in the specified output directory (in Optuna, "docs/source/uttorial"). Sphinx Gallery is kicked implicitly by Sphinx command, i.e., `make html` in the `docs` directory.

**Comment by tktran** [comment]
@crcrpar That was nice of you to fix the formatting errors and add the note, thank you. Your changes are pushed verbatim.

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    def prune(self, study: "optuna.study.Study", trial: "optuna.trial.FrozenTrial") -> bool:
+        # Get the latest score reported from this trial
+        step = trial.last_step
+
+        if step:  # trial.last_step == None when no scores have been reported yet
+            this_score = trial.intermediate_values[step]
+
+            # Get scores from other trials in the study reported at the same step
+            all_trials = study.get_trials(deepcopy=False)
+            completed_trials = [t for t in all_trials if t.state == TrialState.COMPLETE]
```
Comment (hvy) [inline tutorial/20_recipes/006_user_defined_pruner.py:81]: We recently introduced a second argument to `Study.get_trials` to filter on states. You can save some compute time and reduce code. It's definitely optional since we still have code all over Optuna like the one you provided.
```suggestion
            completed_trials = study.get_trials(deepcopy=False, states=(TrialState.COMPLETE,))
```
_Suggestion is not verified._

**Comment by HideakiImamura** [comment]
@hvy @crcrpar Let me assign you to review this PR.

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+about to return True (indicating pruning).
+
+It may be important to note that the SGDClassifier score, as it is evaluated on
+a holdout set, decreases with enough training steps due to overfitting. This
+means that a trial could be pruned even if it had a favorable (high) value on a
+previous training set. After pruning, Optuna will take the intermediate value
+last reported as the value of the trial.
+
+"""
+
```
Comment (crcrpar) [inline tutorial/20_recipes/006_user_defined_pruner.py:56]:
```suggestion


```

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        trial.report(score, step)
+
+        if trial.should_prune():
+            raise optuna.TrialPruned()
+
+    return score
+
+
+
```
Comment (crcrpar) [inline tutorial/20_recipes/006_user_defined_pruner.py:126]:
```suggestion
```

**Review by crcrpar — COMMENTED** [review]
Sorry for having been inactive, let's have this file conform to the style guide and move forward :)

File: tutorial/20_recipes/006_user_defined_pruner.py (hunk 5 above)
Comment (tktran) [reply tutorial/20_recipes/006_user_defined_pruner.py]: Thank you, I committed your suggestions. I am not sure why the CI checks are still failing (the 'black' check specifically), however.

File: tutorial/20_recipes/006_user_defined_pruner.py (hunk 3 above)
Comment (tktran) [reply tutorial/20_recipes/006_user_defined_pruner.py]: @hvy Hello, sorry I haven't worked on this PR in a while. But I noticed that the CI checks fail to run the .py program generated from this PR, since the version of Optuna that they use does not use the updated get_trials() signature with the states= argument.

Comment (hvy) [reply tutorial/20_recipes/006_user_defined_pruner.py]: Thanks, noticed that you managed to fix this by merging the master. (c.f. https://circleci.com/api/v1.1/project/github/optuna/optuna/65256/output/104/0?file=true&allocation-id=5ffe921a1611687798db53ce-0-build%2F708C4094)

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+###################################################################################################
+# Lastly, let's confirm the implementation is correct with the simple hyperparameter optimization.
+
+def objective(trial):
+    iris = load_iris()
+    classes = np.unique(iris.target)
+    X_train, X_valid, y_train, y_valid = train_test_split(
+        iris.data, iris.target, train_size=100, test_size=50, random_state=0)
```
Comment (hvy) [inline tutorial/20_recipes/006_user_defined_pruner.py:106]: Could you fix the CI failure?
```suggestion
        iris.data, iris.target, train_size=100, test_size=50, random_state=0
    )
```

**Review by hvy — COMMENTED** [review]
Thanks for the updates. I left a small comment regarding the style check but otherwise looks good!

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+###################################################################################################
+# Lastly, let's confirm the implementation is correct with the simple hyperparameter optimization.
+
+def objective(trial):
+    iris = load_iris()
+    classes = np.unique(iris.target)
+    X_train, X_valid, y_train, y_valid = train_test_split(
+        iris.data, iris.target, train_size=100, test_size=50, random_state=0
+        )
```
Comment (hvy) [inline tutorial/20_recipes/006_user_defined_pruner.py:107]:
```suggestion
    )
```

**Review by hvy — CHANGES_REQUESTED** [review]
You seemingly committed a change to overwrite the style fix. Could you fix it again?

**Comment by tktran** [comment]
@hvy Thank you for your help. I was having confusion with the CI checks, since they were failing whether I indented the closing parentheses of that line or not. The "black" code formatter was also giving me the "Oh no!" failure message in the middle of the code block, even though the formatting issue was with the spacing before the function signature. I also realized that one way of making sure the CI tests would pass was to run "black ." locally before committing. (I hope my story helps any newbie trying to make contributions to this project in the future.)

Everything finally passes the checks now.

**Review by hvy — APPROVED** [review]
Sorry for the delayed response, and again thanks for your contribution. LGTM!

I'll approve since @crcrpar probably wants to take another look. I noticed that there are some minor markups that could be improved (e.g. not using italics for `True` and `False` but proper refs and similarly the `SGDClassifier`). This is another details but we could also make sure that the pruner works for minimization as well by taking into account `study.direction`. I could incorporate those changes in a follow up PR if you'd be busy.

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
+"""
+.. _user_defined_pruner:
+
+User-Defined Pruner
+====================
```
Comment (crcrpar) [inline tutorial/20_recipes/006_user_defined_pruner.py:5]: super nit-picking
```suggestion
User-Defined Pruner
===================
```

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+-----------------------------
+
+The :func:`~optuna.study.create_study` constructor takes, as an optional
+argument, a pruner inheriting from :class:`~optuna.pruners.BasePruner`. The
+pruner should implement the abstract method
+:meth:`~optuna.pruners.BaseSampler.prune`, which takes arguments for the
+associated :class:`~optuna.study.Study` and :class:`~optuna.trial.Trial` and
+returns a boolean value: `True` if the trial should be pruned and `False`
+otherwise. Using the Study and Trial objects, you can access all other trials
+through the :meth:`study.get_trial` method and, and from a trial, its reported
```
Comment (crcrpar) [inline tutorial/20_recipes/006_user_defined_pruner.py:23]:
```suggestion
through the :func:`optuna.study.Study.get_trial` method and, and from a trial, its reported
```

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+An Example: Implementing ``LastPlacePruner``
+--------------------------------------------
+
+We aim to optimize the ``loss`` and ``alpha`` hyperparameters for a stochastic
+gradient descent classifier (``SGDClassifier``) run on the sklearn iris dataset. We
+implement a pruner which terminates a trial at a certain step if it is in last
+place compared to completed trials at the same step. We begin considering
+pruning after a "warmup" of 1 training step and 5 completed trials. For
+demonstration purposes, we print() a diagnostic message from prune() when it is
+about to return True (indicating pruning).
```
Comment (crcrpar) [inline tutorial/20_recipes/006_user_defined_pruner.py:47]:
```suggestion
about to return :obj:`True` (indicating pruning).
```

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+Overview of Pruning Interface
+-----------------------------
+
+The :func:`~optuna.study.create_study` constructor takes, as an optional
+argument, a pruner inheriting from :class:`~optuna.pruners.BasePruner`. The
+pruner should implement the abstract method
+:meth:`~optuna.pruners.BaseSampler.prune`, which takes arguments for the
+associated :class:`~optuna.study.Study` and :class:`~optuna.trial.Trial` and
+returns a boolean value: `True` if the trial should be pruned and `False`
```
Comment (crcrpar) [inline tutorial/20_recipes/006_user_defined_pruner.py:21]:
```suggestion
returns a boolean value: :obj:`True` if the trial should be pruned and :obj:`False`
```

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+describe how to implement your own pruner, i.e., a custom strategy for
+determining when to stop a trial.
+
+Overview of Pruning Interface
+-----------------------------
+
+The :func:`~optuna.study.create_study` constructor takes, as an optional
+argument, a pruner inheriting from :class:`~optuna.pruners.BasePruner`. The
+pruner should implement the abstract method
+:meth:`~optuna.pruners.BaseSampler.prune`, which takes arguments for the
```
Comment (crcrpar) [inline tutorial/20_recipes/006_user_defined_pruner.py:19]:
```suggestion
:func:`~optuna.pruners.BaseSampler.prune`, which takes arguments for the
```

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+associated :class:`~optuna.study.Study` and :class:`~optuna.trial.Trial` and
+returns a boolean value: `True` if the trial should be pruned and `False`
+otherwise. Using the Study and Trial objects, you can access all other trials
+through the :meth:`study.get_trial` method and, and from a trial, its reported
+intermediate values through the `trial.intermediate_values` attribute (a
```
Comment (crcrpar) [inline tutorial/20_recipes/006_user_defined_pruner.py:24]:
```suggestion
intermediate values through the :attribute:`~optuna.trial.FrozenTrial.intermediate_values` (a
```

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+dictionary which maps an integer `step` to a float value).
```
Comment (crcrpar) [inline tutorial/20_recipes/006_user_defined_pruner.py:25]:
```suggestion
dictionary which maps an integer ``step`` to a float value).
```

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+
+An Example: Implementing ``LastPlacePruner``
+--------------------------------------------
+
+We aim to optimize the ``loss`` and ``alpha`` hyperparameters for a stochastic
+gradient descent classifier (``SGDClassifier``) run on the sklearn iris dataset. We
+implement a pruner which terminates a trial at a certain step if it is in last
+place compared to completed trials at the same step. We begin considering
+pruning after a "warmup" of 1 training step and 5 completed trials. For
+demonstration purposes, we print() a diagnostic message from prune() when it is
```
Comment (crcrpar) [inline tutorial/20_recipes/006_user_defined_pruner.py:46]: print() and prune() can be better rendered in built documentation.

File: tutorial/20_recipes/006_user_defined_pruner.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+about to return True (indicating pruning).
+
+It may be important to note that the SGDClassifier score, as it is evaluated on
```
Comment (crcrpar) [inline tutorial/20_recipes/006_user_defined_pruner.py:49]:
```suggestion
It may be important to note that the ``SGDClassifier`` score, as it is evaluated on
```

**Review by crcrpar — APPROVED** [review]
@tktran First of all, apologies for my being late, I had some paperwork.

Thank you for your awesome work, overall looks great except for trivial visual things (I want to emphasize that the built documentation looks already really cool).
As this pull request has been running long, I'd like to merge this first and then improve the comments I dropped (note that my suggestions may not work as expected).

Let me know whether you kindly will do the visual things or not. :)

File: tutorial/20_recipes/006_user_defined_pruner.py (hunk 16 above)
Comment (tktran) [reply tutorial/20_recipes/006_user_defined_pruner.py:49]: Hello, thank you for your suggestions. I will follow your suggested course of action - wait for you to merge, and then I can incorporate these additional visual corrections after testing them myself.

**Comment by tktran** [comment]
Hello, thank you for your suggestions @crcrpar and @hvy.

@crcrpar I will follow your suggested course of action - wait for you to merge, and then I can incorporate the suggested visual corrections after testing them myself.

@hvy In the best interest of the project, I think that I should leave the additional improvements to you and your team, as I am not sure of the direction you want to take them. As for me, I will spend some time tackling another issue to prepare myself for a GSoC proposal.
